### PR TITLE
New thread priority names

### LINKED ISSUE #2747: Inherited task priorities break certain DAG optimizations
Debugging problems in performance with our matrix code has revealed that the following check
https://github.com/STEllAR-GROUP/hpx/blob/d114fc3011c70fd2e70f146280ca943f86080b38/hpx/runtime/threads/detail/create_thread.hpp#L76-L77
essentially makes it impossible to create complex DAGs that spawn sub tasks from parent tasks, when only some of the child tasks are lower priority.
The effect of this check is to make all child tasks HP. In our matrix solver the first row/col block on each diagonal is high priority and the first row/col is handled first, so all tasks are children of it and in turn, all become HP.

Two possible solutions spring to mind

1. Remove the check so that child tasks are not promoted
2. Add a new priority between `thread_priority_normal` and `thread_priority_critical` which would be `thread_priority_high` and would give the current task high priority, but not child tasks.
3. Rename one policy to `thread_priority_high_recursive` to make it more clear that this is passed on to children

I suspect that I should have been using `thread_priority_boost` instead of `thread_priority_critical` right from the start, but had not appreciated the distinction between them regarding child tasks. Either way, some renaming of priorities might make this more clear.

### PR DESCRIPTION
**Pull request by hkaiser**
Introducing explicit high priority and high priority (recursive) as thread priorities, deprecating `thread_priority_critical` (replaced by `thread_priority_high_recursive`).

This fixes #2747

@biddisco Please review.

### CODE REVIEW

**Comment by hkaiser** [comment]
From the comments you made on #2602 I take it that this can go ahead now, can't it?

**Comment by biddisco** [comment]
My impression so far is that this is a big improvement and it can be merged if you're in a hurry.

I wanted to play some more just to check all corner cases.

**Comment by hkaiser** [comment]
Ok, we'll wait for your verdict, then.

File: hpx/runtime/threads/detail/create_work.hpp [hunk 1]
```diff
@@ -81,12 +81,16 @@ namespace hpx { namespace threads { namespace detail
         // Pass critical priority from parent to child.
         if (self)
         {
-            if (thread_priority_critical == threads::get_self_id()->get_priority())
-                data.priority = thread_priority_critical;
+            if (thread_priority_high_recursive ==
+                threads::get_self_id()->get_priority())
+            {
+                data.priority = thread_priority_high_recursive;
```
Comment (biddisco) [inline hpx/runtime/threads/detail/create_work.hpp]: Should this check that the promoted thread is "default" or normal. If I have a child task of a high_riority_recursive and I launch it as low priority, I think it should go low and not be promoted to high.

Comment (hkaiser) [inline hpx/runtime/threads/detail/create_work.hpp]: Right, this should additionally check whether `data.priority == thread_priority_default`. Which also means that we should make `thread_priority_default` and `thread_priority_normal` distinct in terms of enum values represented (just treat them in similar ways).

Comment (hkaiser) [inline hpx/runtime/threads/detail/create_work.hpp]: Excellent catch, btw!

**Comment by hkaiser** [comment]
@biddisco: Would you be able to check one more time whether my latest changes are fine with the application you're using to test, please?

**Comment by biddisco** [comment]
Seems to be working ok with the latest changes. I'm happy to see this merged ASAP.

**Pull request merged**